### PR TITLE
test_execution_trace.py: Use instantiate_device_type_tests to run GPU tests on HPU as well

### PR DESCRIPTION
**Pull request by amathewc**
**MOTIVATION**

We recently integrated support for Intel Gaudi devices (identified as 'hpu') into the common_device_type framework via the pull request at https://github.com/pytorch/pytorch/pull/126970. This integration allows tests to be automatically instantiated for Gaudi devices upon loading the relevant library. Building on this development, the current pull request extends the utility of these hooks by adapting selected CUDA tests to operate on Gaudi devices. Additionally, we have confirmed that these modifications do not interfere with the existing tests on CUDA devices. 

**CHANGES**

- Add support for HPU devices within the payload function.
- Use instantiate_device_type_tests with targeted attributes to generate device-specific test instances.
- Expand the supported_activities() function to include checks for torch.profiler.ProfilerActivity.HPU.
- Apply skipIfHPU decorator to bypass tests that are not yet compatible with HPU devices.

### CODE REVIEW

**Comment by amathewc** [comment]
Hi @sraikund16 :
Just wanted to remind you about this PR. Please let me know if you need any additional information or changes from my side.
Thanks!

**Comment by aaronenyeshi** [comment]
Could this diff be stacked on top of this refactor? https://github.com/pytorch/pytorch/pull/134316

**Comment by amathewc** [comment]
> Could this diff be stacked on top of this refactor? #134316

@aaronenyeshi : It seems like the PR mentioned (https://github.com/pytorch/pytorch/pull/134316 ) provides support of in-tree devices. Since Intel Gaudi (HPU) is an out-of-tree device, we needed a different approach . Please refer to this https://github.com/pytorch/pytorch/pull/126970 as well. 
Adding @jgong5  and @ankurneog for further comments.

**Comment by ankurneog** [comment]
@aaronenyeshi  : i agree with @amathewc  , we can give the flexibility per file/module. Each device based on the its capability can include itself per module /file  (eg : https://github.com/pytorch/pytorch/blob/1011e0ae9804baca9ef06abfeb6903f15459d346/test/dynamo/test_model_output.py#L348)
we should merge both the changes inline so that its useful for out-of-tree devices such as Intel Gaudi.

**Comment by aaronenyeshi** [comment]
Sounds good, cc @sanrise , @briancoutinho , @shengfukevin - please take a look. The changes are related to the execution trace test.

File: test/profiler/test_execution_trace.py [hunk 1]
```diff
@@ -117,14 +120,16 @@ def get_kineto_rf_ids(self, events: List[Json]) -> List[int]:
...
+    @skipIfHpu
+    def test_execution_trace_with_kineto(self, device):
         trace_called_num = 0
 
         def trace_handler(p):
             nonlocal trace_called_num
             trace_called_num += 1
 
-        use_cuda = torch.profiler.ProfilerActivity.CUDA in supported_activities()
+        #use_cuda = torch.profiler.ProfilerActivity.CUDA in supported_activities()
```
Comment (briancoutinho) [inline test/profiler/test_execution_trace.py:131]: could we remove this line

File: test/profiler/test_execution_trace.py [hunk 2]
```diff
@@ -275,8 +283,9 @@ def fn(a, b, c):
                         assert len(n["outputs"]["values"]) == 0
         assert found_captured_triton_kernel_node
 
-    def test_execution_trace_start_stop(self):
-        use_cuda = torch.profiler.ProfilerActivity.CUDA in supported_activities()
+    def test_execution_trace_start_stop(self, device):
+        #use_cuda = torch.profiler.ProfilerActivity.CUDA in supported_activities()
```
Comment (briancoutinho) [inline test/profiler/test_execution_trace.py:287]: please remove the extra line

File: test/profiler/test_execution_trace.py [hunk 3]
```diff
@@ -310,8 +319,9 @@ def test_execution_trace_start_stop(self):
         assert found_root_node
         assert loop_count == expected_loop_events
 
-    def test_execution_trace_repeat_in_loop(self):
-        use_cuda = torch.profiler.ProfilerActivity.CUDA in supported_activities()
+    def test_execution_trace_repeat_in_loop(self, device):
+        #use_cuda = torch.profiler.ProfilerActivity.CUDA in supported_activities()
```
Comment (briancoutinho) [inline test/profiler/test_execution_trace.py:323]: same

**Comment by briancoutinho** [comment]
Looks great, just some minor suggestions. @amathewc  can you rebase on viablestrict branch, that should ensure all the checks will be passing.

**Review by briancoutinho — APPROVED** [review]

File: test/profiler/test_execution_trace.py (hunk 3 above)
Comment (amathewc) [reply test/profiler/test_execution_trace.py]: Updated.

File: test/profiler/test_execution_trace.py (hunk 2 above)
Comment (amathewc) [reply test/profiler/test_execution_trace.py]: Removed the lines.

**Comment by amathewc** [comment]
@aaronenyeshi , @sanrise , @briancoutinho , @shengfukevin : Removed the extra lines as per review comments.

**Review by aaronenyeshi — APPROVED** [review]

**Comment by amathewc** [comment]
@shengfukevin , @sanrise , @sraikund16  : Any further changes required from our side for merging this PR ?

**Comment by sraikund16** [comment]
@amathewc Can you rebase/merge using git and push to the branch to start testing

**Comment by amathewc** [comment]
@aaronenyeshi , @sanrise , @briancoutinho , @shengfukevin, @sraikund16 : could you help in rebase/merge this ? Looks like I do not have the permissions. 
The CI failures are in unrelated files.

**Comment by amathewc** [comment]
@aaronenyeshi , @sanrise , @briancoutinho , @shengfukevin, @sraikund16 : Could you help in merging this ? The CI failures seem to be in unrelated files.

**Comment by amathewc** [comment]
@aaronenyeshi , @sanrise , @briancoutinho , @shengfukevin, @sraikund16 , @jgong5  : Could you help in merging this ? I don't seem to be authorized to merge this. Have fixed all lint related issues.

**Pull request closed without merging**